test/blackbox-tests/test-cases/pkg/convert-opam-commands.t

```diff
@@ -0,0 +1,69 @@
+Helper shell function that generates an opam file for a package:
+  $ mkpkg() {
+  >   name=$1
+  >   mkdir -p mock-opam-repository/packages/$name/$name.0.0.1
+  >   cat >mock-opam-repository/packages/$name/$name.0.0.1/opam
+  > }
+
+Helper shell function to generate a dune-project file and generate lockdir:
+  $ solve_project() {
+  >   cat >dune-project
+  >   dune pkg lock --opam-repository-path=mock-opam-repository
+  > }
+
+Generate a mock opam repository
+  $ mkdir -p mock-opam-repository
+  $ cat >mock-opam-repository/repo <<EOF
+  > opam-version: "2.0"
+  > EOF
+
+  $ mkpkg standard-dune <<EOF
+  > opam-version: "2.0"
+  > build: [
+  >   ["dune" "subst"] {dev}
+  >   [
+  >     "dune"
+  >     "build"
+  >     "-p"
+  >     name
+  >     "-j"
+  >     jobs
+  >     "@install"
+  >     "@runtest" {with-test}
+  >     "@doc" {with-doc}
+  >   ]
+  > ]
+  > install: [ make "install" ]
+  > EOF
+
+  $ mkpkg with-unknown-variable <<EOF
+  > opam-version: "2.0"
+  > build: [ fake "install" ]
+  > EOF
+
+  $ solve_project <<EOF
+  > (lang dune 3.8)
+  > (package
+  >  (name x)
+  >  (depends standard-dune))
+  > EOF
+  Solution for dune.lock:
+  standard-dune.0.0.1
+  
+
+  $ cat dune.lock/standard-dune.pkg
+  (version 0.0.1)
+  (install (progn (system "%{make} install")))
+  (build (progn (system "dune subst") (system "dune build -p %{name} -j %{jobs} @install @runtest @doc")))
```

Comment:
Is using `system` correct? I imagine it should be `(run dune subst)`.

Reply:
I thought `(system "dune subst")` and `(run dune subst)` are equivalent. According to the dune docs:

> Execute a command using the system shell: sh on Unix and cmd on Windows.
> Example:
> `(system "command arg1 arg2")`

I deliberately chose to use `system` instead of `run` as opam commands might include `&&` or similar operators and I don't think it's possible to do `(run make && make install)` say whereas `(system "make && make install")` should work.

Reply:
Well you can turn && into a progn with a bit of parsing.

Reply:
Sure, but for this initial implementation let's keep things as simple as possible. Parsing commands can get complicated since these can be arbitrary commands with "&&" possibly appearing in a string, etc. What's the benefit of representing commands with `Run` actions over `System` actions?

Reply:
I would also suggest `system` (or `bash` even), given we probably don't want to reimplement a whole parser for shell syntax for now, especially given we support shell commands already.

Reply:
I see. Opam is using shell then we should copy what it's doing. However, is opam really using `sh` or `bash`? Could you double check that.

Reply:
I did some digging on how OPAM runs commands. `OpamAction.build_package` collects the build commands and converts them eventually into some `OpamProcess.Job` structure and if I see it correctly to evaluate one of these it does call  `Unix.create_process_env` in the end. 

So it doesn't seem to involve a shell, thus presumably `&&` and redirections wouldn't be valid, a package would need to wrap it in `sh -c`.

By the way, we have our own runner for `OpamProcess.Job` that uses Fibers, so maybe the part that collects commands in OPAM could be leveraged. Don't even necessarily need to run them, we can also just unpack a job into its commands again and write them into the lockfile, since the `OpamProcess.Job` data structure exposes all the internal parts.

Reply:
Thanks for looking into this @Leonidas-from-XIV. I wasn't expecting it to work that way. I'll see if the way that opam represents commands can be mapped onto dune's `run` action.

Reply:
> By the way, we have our own runner for OpamProcess.Job that uses Fibers, so maybe the part that collects commands in OPAM could be leveraged. Don't even necessarily need to run them, we can also just unpack a job into its commands again and write them into the lockfile, since the OpamProcess.Job data structure exposes all the internal parts.

We need to work with actions here otherwise the build system will not offer incremental building.

> I wasn't expecting it to work that way. I'll see if the way that opam represents commands can be mapped onto dune's run action.

FYI, I expected that it would map quite naturally. However, if it doesn't, feel free to add new constructs to the action language to match opam's semantics.